File: Java/src/main/java/com/semantria/JobIdFeatureTestApp.java
package com.semantria;

import com.semantria.interfaces.ISerializer;
import com.semantria.mapping.Document;
import com.semantria.mapping.output.DocAnalyticData;
import com.semantria.serializer.JsonSerializer;
import org.apache.commons.lang.StringUtils;

import java.io.*;
import java.util.*;

public class JobIdFeatureTestApp {
    private static final int TIMEOUT_BEFORE_GETTING_RESPONSE = 10000; //in millisec

    public static void main(String args[]) {
        final String key = "";
        final String secret = "";
        List<String> initialTexts = new ArrayList<String>();

        System.out.println("Semantria service demo.");

        File file = new File( DetailedModeTestApp.class.getProtectionDomain().getCodeSource().getLocation().toString().replace("/target/classes", "/src/main/java/" + JobIdFeatureTestApp.class.getPackage().getName().replace(".","/")).replace("file:/", "")+ "/source.txt");

        if( !file.exists() ) {
            System.out.println("Source file isn't available.");
            return;
        }

        //Reads dataset from the source file
        try
        {
            FileInputStream fstream = new FileInputStream(file);
            DataInputStream in = new DataInputStream(fstream);
            BufferedReader br = new BufferedReader(new InputStreamReader(in));
            String strLine;

            while ((strLine = br.readLine()) != null)
            {
                if (StringUtils.isEmpty(strLine) || strLine.length() < 3)
                {
                    continue;
                }

                initialTexts.add(strLine);
            }
            in.close();
        }
        catch (Exception e){
            System.err.println("Error: " + e.getMessage());
        }

        ISerializer serializer = new JsonSerializer();
        Session session = Session.createSession(key, secret, serializer, true);
        session.setCallbackHandler(new CallbackHandler());

        // null - send every single document separately
        // false - send uniqueJobIdCount batches
        // true - send all documents in single batch
        Boolean dataSendingMode = true;
        int uniqueJobIdCount = 4;
        Map<Integer, Integer> jobIds = new HashMap<Integer, Integer>();
        Map<Integer, List<Document>> documents = new HashMap<Integer, List<Document>>();
        Random randomGenerator = new Random();
        int max = 10000000;
        int min = 0;

        //Generates N unique jobId values
        for (int index = 0; index < uniqueJobIdCount; index++) {
            Integer id = randomGenerator.nextInt((max - min) + 1) + min;

            jobIds.put(id, 0);
            documents.put(id, new ArrayList<Document>());
        }

        List<Integer> jobIdsKeys = new ArrayList<Integer>(jobIds.keySet());
        for (int i = 0; i < initialTexts.size(); i++) {
            String docText = initialTexts.get(i);

            // Creates a sample document which need to be processed on Semantria
            Integer jobId = jobIdsKeys.get(randomGenerator.nextInt(jobIds.size()));

            jobIds.put(jobId, jobIds.get(jobId) + 1);
            List<Document> docs = documents.get(jobId);
            String docId = String.valueOf(randomGenerator.nextInt((max - min) + 1) + min);
            Document doc = new Document(docId, docText);
            doc.setJobId(String.valueOf(jobId));
            docs.add(doc);
        }

        System.out.println("Queues document for processing on Semantria service...");
        for (Map.Entry<Integer, Integer> entry : jobIds.entrySet())  {
            System.out.println(entry.getValue() + " documents for " + entry.getKey() + " jobId");
        }

        if (dataSendingMode == null) {
            for (Map.Entry<Integer, List<Document>> entry : documents.entrySet())  {
                List<Document> docsForJob = entry.getValue();
                for (Document doc : docsForJob) {
                    // Queues document for processing on Semantria service
                    session.queueDocument(doc);
                }
                System.out.println(docsForJob.size() + " documents queued for " + entry.getKey() + " job ID");
            }
        } else if (!dataSendingMode) {
            for (Map.Entry<Integer, List<Document>> entry : documents.entrySet())  {
                // Queues document for processing on Semantria service
                session.queueBatch(entry.getValue());
                System.out.println(entry.getValue().size() + " documents queued for " + entry.getKey() + " job ID");
            }
        } else {
            List<Document> batch = new ArrayList<Document>();

            for (Map.Entry<Integer, List<Document>> entry : documents.entrySet())  {
                batch.addAll(entry.getValue());
            }
            session.queueBatch(batch);
            System.out.println(batch.size() + " documents queued in single batch");
        }


        // Receive response
        boolean allJobIdsProcessed;
        try {
            do {
                Thread.currentThread().sleep(TIMEOUT_BEFORE_GETTING_RESPONSE);

                System.out.println("\nRetrieving processed results...");
                allJobIdsProcessed = true;

                // Requests processed results from Semantria service
                for (Map.Entry<Integer, Integer> entry : jobIds.entrySet()) {
                    int count = 0;

                    if (entry.getValue() > 0) {
                        allJobIdsProcessed = false;
                        List<DocAnalyticData> processedDocuments = session.getProcessedDocumentsByJobId(String.valueOf(entry.getKey()));
                        if (processedDocuments.size() > 0) {
                            jobIds.put(entry.getKey(), entry.getValue() - processedDocuments.size());
                            count += processedDocuments.size();
                        }

                        if (count > 0) {
                            System.out.println(count + " documents received for " + entry.getKey() + " job ID");
                        }
                    }
                }

                if (allJobIdsProcessed) {
                    System.out.println("\nDone. All documents received.");
                }
            } while (!allJobIdsProcessed);
        } catch(Exception e){
            System.out.println(e.getMessage());
        }
    }
}

File: Node.js/examples/JobIdFeatureTestApp.js
var SemantriaSession = require("../").Session;
var promise = require('promise');
var config = require('../test-config');
try { config = require('../test-config.override') } catch(e) {}

var consumerKey = config.consumerKey,
	consumerSecret = config.consumerSecret,
	/**
	 * @type {(boolean|null)} dataSendingMode
	 *   null - send every single document separately
	 *   false - send uniqueJobIdCount batches
	 *   true - send all documents in single batch
	 */
	dataSendingMode = false,
	appConfigurationId = null,
	appConfigurationName = "JobIdFeatureModeTestApp Configuration",
	jobIds = {},
	uniqueJobIdCount = 4,
	documents = {},
	SemantriaActiveSession = new SemantriaSession(consumerKey, consumerSecret, "myApp");

if (config.apiHost) {
	SemantriaActiveSession.API_HOST = config.apiHost
}

console.log("Semantria JobId feature demo.");

//Generates N unique jobId values
for (var index = 0; index < uniqueJobIdCount; index++) {
	var id = Math.floor(Math.random() * 10000000);

	jobIds[id] = 0;
	documents[id] = [];
}

var initialTexts = require('./source.json');
for (var i=0, item; item=initialTexts[i]; i++) {
	// Creates a sample document which need to be processed on Semantria
	var id = Math.floor(Math.random() * 10000000),
		jobId = randomProperty(jobIds);

	jobIds[jobId]++;
	documents[jobId].push({
		id: '' + id,
		text: item,
		job_id: jobId
	});
}

console.log("Queues document for processing on Semantria service...");
for (var jobId in jobIds) {
	console.log(jobIds[jobId] + " documents for " + jobId + " jobId");
}

//get or create test app configuration
SemantriaActiveSession.getConfigurations(true)
.then(function(configurations) {
	for (var i=0; i<configurations.length; i++) {
		if (configurations[i].name == appConfigurationName) {
			return promise.resolve([configurations[i]]);
		}
	}
	return SemantriaActiveSession.addConfigurations([{
		name: appConfigurationName,
		is_primary: false,
		auto_response: false,
		language: "English"
	}], true);
})
.then(function (configuration) {
	appConfigurationId = configuration[0].id;

	//queued documents
	if (dataSendingMode === null) return queueDocumentSingle();
	if (dataSendingMode === false) return queueDocumentBatches();
	if (dataSendingMode === true) return queueDocumentSingleBatch();

	return new promise.reject("incorrect dataSendingMode");
})
.then(function(){
	//Wait while Semantria process queued document
	return new promise(function(resolve, reject) {
		var analyticData = [];
		var wait_fn = function () {
			console.log("Retrieving your processed results...");

			var jobId, jobs = [];
			for (jobId in jobIds) {
				jobs.push(SemantriaActiveSession.getProcessedDocumentsByJobId(jobId, true));
			}


			promise.all(jobs)
			.then(function(result) {

				var i = 0;
				for (jobId in jobIds) {
					var processedDocuments = result[i];
					if (processedDocuments && processedDocuments.length) {
						jobIds[jobId] -= processedDocuments.length;
						console.log("%d documents received for %s job ID", processedDocuments.length, jobId);
						analyticData.concat(processedDocuments);
					}
					i++;
				}

				var waits = [];
				for (jobId in jobIds) {
					if (jobIds[jobId] > 0) {
						waits.push('jod#'+jobId+' - '+ jobIds[jobId]);
					}
				}

				if (waits.length) {
					//console.log(" wait: " + waits.join(', '));
					setTimeout(wait_fn, 5000);
				} else {
					resolve(analyticData);
				}
			})
			.catch(reject);
		}
		setTimeout(wait_fn, 5000);
	});

})
.then(function(analyticData){
	console.log("Done. All documents received.");
})
.catch(function(err) {
	console.log("JobIdFeatureModeTestApp faild\n");
	console.log(err);
	console.log(err.stack);
})
.then(function() {
	if (!appConfigurationId) return;
	return SemantriaActiveSession.removeConfigurations([appConfigurationId], true);
});

function queueDocumentSingle() {
	var jobs = [];
	for (var jobId in documents) {
		var docsForJob = documents[jobId];
		var docs = []
		for (index in docsForJob) {
			// Queues document for processing on Semantria service
			docs.push(SemantriaActiveSession.queueDocument(docsForJob[index], appConfigurationId, true));
		}
		(function(count, jobId) {
			var job = promise.all(docs).then(function() {
				console.log("%d documents queued for %s job ID", count, jobId);
			});
			jobs.push(job);
		})(Object.keys(docsForJob).length, jobId);
	}

	return promise.all(jobs);
}

function queueDocumentBatches() {
	var jobs = [];
	for (var jobId in documents) {
		// Queues document for processing on Semantria service
		(function (jobId) {
			SemantriaActiveSession.queueBatchOfDocuments(documents[jobId], appConfigurationId, true).then(function () {
				console.log("%d documents queued for %s job ID", Object.keys(documents[jobId]).length, jobId);
			});
		})(jobId);
	}

	return promise.all(jobs);
}

function queueDocumentSingleBatch() {
	var batch = [];

	for (var jobId in documents) {
		batch = batch.concat(documents[jobId]);
	}
	return SemantriaActiveSession.queueBatchOfDocuments(batch, appConfigurationId, true).then(function() {
		console.log("%d documents queued in single batch", batch.length);

	});
}

function randomProperty(obj) {
	var keys = Object.keys(obj)
	return keys[ keys.length * Math.random() << 0];
}
